### PR TITLE
Security check

### PR DESCRIPTION
**Pull request by Avielyo10**
# Description

Fix all of security alerts on our project & implement a mechanism to catch them in the future

Fixes security possible risks

## Type of change

What is it?

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I run `go fmt ./...` to check that my code is properly formatted
- [x] I run `go vet ./...` to check that my code is free of common Go style mistakes

### CODE REVIEW

**Review by biancarosa — COMMENTED** [review]
Can you keep the log messages clean of variables and do something like? 

services.Log.WithField("error", &err).Error("Error while trying to encode")

**Comment by Avielyo10** [comment]
@biancarosa This will take time but yeah sure

**Comment by Avielyo10** [comment]
@biancarosa changes pushed

File: pkg/clients/imagebuilder/client.go [hunk 1]
```diff
@@ -390,7 +392,9 @@ func (c *Client) GetMetadata(image *models.Image) (*models.Image, error) {
 	}
 
 	var metadata Metadata
-	json.Unmarshal(respBody, &metadata)
+	if err := json.Unmarshal(respBody, &metadata); err != nil {
+		c.log.Error("Error while trying to unmarshal ", metadata)
```
Comment (biancarosa) [inline pkg/clients/imagebuilder/client.go:396]: Can we return this error?

File: pkg/clients/inventory/client.go [hunk 2]
```diff
@@ -191,7 +191,9 @@ func (c *Client) ReturnDevicesByID(deviceID string) (Response, error) {
 		return Response{}, fmt.Errorf("error requesting InventoryResponse, got status code %d and body %s", res.StatusCode, body)
 	}
 	var inventory Response
-	json.Unmarshal([]byte(body), &inventory)
+	if err := json.Unmarshal([]byte(body), &inventory); err != nil {
+		log.Error("Error while trying to unmarshal ", &inventory)
```
Comment (biancarosa) [inline pkg/clients/inventory/client.go:195]:
```suggestion
		c.log.Error("Error while trying to unmarshal ", &inventory)
```

Comment (biancarosa) [reply pkg/clients/inventory/client.go:195]: And maybe add a return too

File: pkg/clients/inventory/client.go [hunk 3]
```diff
@@ -233,6 +235,8 @@ func (c *Client) ReturnDevicesByTag(tag string) (Response, error) {
...
```
Comment (biancarosa) [inline pkg/clients/inventory/client.go:240]:
```suggestion
		c.log.Error("Error while trying to unmarshal ", &inventory)
		return Response{}, err
```

File: pkg/clients/playbookdispatcher/client.go [hunk 4]
```diff
@@ -87,6 +89,8 @@ func (c *Client) ExecuteDispatcher(payload DispatcherPayload) ([]Response, error
 	}
 
 	var playbookResponse []Response
-	json.Unmarshal([]byte(body), &playbookResponse)
+	if err := json.Unmarshal([]byte(body), &playbookResponse); err != nil {
+		c.log.Error("Error while trying to unmarshal ", &playbookResponse)
```
Comment (biancarosa) [inline pkg/clients/playbookdispatcher/client.go:93]:
```suggestion
		c.log.Error("Error while trying to unmarshal ", &playbookResponse)
		return nil, err
```

File: pkg/routes/inventory.go [hunk 5]
```diff
@@ -65,7 +65,9 @@ func GetInventory(w http.ResponseWriter, r *http.Request) {
 	InventoryData.Total = inventory.Total
 	InventoryData.Results = results
 
-	json.NewEncoder(w).Encode(InventoryData)
+	if err := json.NewEncoder(w).Encode(InventoryData); err != nil {
+		fmt.Println("Error while trying to encode ", InventoryData)
```
Comment (biancarosa) [inline pkg/routes/inventory.go:69]: Should we return this error?

File: pkg/routes/imagesets.go [hunk 6]
```diff
@@ -295,7 +324,9 @@ func validateFilterParams(next http.Handler) http.Handler {
 			return
 		}
 		w.WriteHeader(http.StatusBadRequest)
-		json.NewEncoder(w).Encode(&errs)
+		if err := json.NewEncoder(w).Encode(&errs); err != nil {
+			log.WithField("error", errs).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/imagesets.go:328]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/imagesets.go [hunk 7]
```diff
@@ -232,15 +251,19 @@ func GetImageSetsByID(w http.ResponseWriter, r *http.Request) {
 		}).Error("Error retrieving account")
 		err := errors.NewBadRequest(err.Error())
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/imagesets.go:255]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/imagesets.go [hunk 8]
```diff
@@ -232,15 +251,19 @@ func GetImageSetsByID(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 	ctx := r.Context()
 	imageSet, ok := ctx.Value(imageSetKey).(*models.ImageSet)
 	if !ok {
 		err := errors.NewBadRequest("Must pass image set id")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/imagesets.go:265]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/imagesets.go [hunk 9]
```diff
@@ -250,7 +273,9 @@ func GetImageSetsByID(w http.ResponseWriter, r *http.Request) {
 	if result.Error != nil {
 		err := errors.NewBadRequest("Error to filter images")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/imagesets.go:277]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 10]
```diff
@@ -43,7 +43,9 @@ func DeviceCtx(next http.Handler) http.Handler {
 		if dc.DeviceUUID == "" {
 			err := errors.NewBadRequest("DeviceUUID must be sent")
 			w.WriteHeader(err.GetStatus())
-			json.NewEncoder(w).Encode(&err)
+			if err := json.NewEncoder(w).Encode(&err); err != nil {
+				log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:47]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 11]
```diff
@@ -62,19 +64,25 @@ func GetUpdateAvailableForDevice(w http.ResponseWriter, r *http.Request) {
...
+			log.WithField("error", result).Error("Error while trying to encode")
+		}
 		return
 	}
 	if _, ok := err.(*services.DeviceNotFoundError); ok {
 		err := errors.NewNotFound("Could not find device")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:76]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 12]
```diff
@@ -62,19 +64,25 @@ func GetUpdateAvailableForDevice(w http.ResponseWriter, r *http.Request) {
 	}
 	result, err := s.DeviceService.GetUpdateAvailableForDeviceByUUID(dc.DeviceUUID)
 	if err == nil {
-		json.NewEncoder(w).Encode(result)
+		if err := json.NewEncoder(w).Encode(result); err != nil {
+			log.WithField("error", result).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:68]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 13]
```diff
@@ -96,13 +106,17 @@ func GetDeviceImageInfo(w http.ResponseWriter, r *http.Request) {
 	}
 	result, err := s.DeviceService.GetDeviceImageInfo(dc.DeviceUUID)
 	if err == nil {
-		json.NewEncoder(w).Encode(result)
+		if err := json.NewEncoder(w).Encode(result); err != nil {
+			log.WithField("error", result).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:110]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 14]
```diff
@@ -111,7 +125,9 @@ func GetDeviceImageInfo(w http.ResponseWriter, r *http.Request) {
 		"statusCode": apierr.GetStatus(),
 		"error":      apierr.Error(),
 	}).Error("Error getting image info for device")
-	json.NewEncoder(w).Encode(&err)
+	if err := json.NewEncoder(w).Encode(&err); err != nil {
+		log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:129]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/services/files/extractor.go [hunk 15]
```diff
@@ -32,24 +34,44 @@ func (f *TARFileExtractor) Extract(rc io.ReadCloser, dst string) error {
...
 	}
 	return nil
 }
+
+func sanitizeExtractPath(filePath string, destination string) (destpath string, err error) {
+	destpath = filepath.Join(destination, filePath)
+	if !strings.HasPrefix(destpath, filepath.Clean(destination)+string(os.PathSeparator)) {
+		err = fmt.Errorf("%s: illegal file path", filePath)
+	}
+	return
```
[the remark this answers is not in the document]
Comment (Avielyo10) [reply pkg/services/files/extractor.go:76]: Yes, will return both.

File: pkg/routes/devices.go [hunk 16]
```diff
@@ -127,19 +143,25 @@ func GetDevice(w http.ResponseWriter, r *http.Request) {
...
+			log.WithField("error", result).Error("Error while trying to encode")
+		}
 		return
 	}
 	if _, ok := err.(*services.ImageNotFoundError); ok {
 		err := errors.NewNotFound("Could not find image")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:155]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 17]
```diff
@@ -148,5 +170,7 @@ func GetDevice(w http.ResponseWriter, r *http.Request) {
 		"statusCode": apierr.GetStatus(),
 		"error":      apierr.Error(),
 	}).Error("Error retrieving updates for device")
-	json.NewEncoder(w).Encode(&err)
+	if err := json.NewEncoder(w).Encode(&err); err != nil {
+		log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:174]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 18]
```diff
@@ -127,19 +143,25 @@ func GetDevice(w http.ResponseWriter, r *http.Request) {
 	}
 	result, err := s.DeviceService.GetDeviceDetails(dc.DeviceUUID)
 	if err == nil {
-		json.NewEncoder(w).Encode(result)
+		if err := json.NewEncoder(w).Encode(result); err != nil {
+			log.WithField("error", result).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:147]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

File: pkg/routes/devices.go [hunk 19]
```diff
@@ -62,19 +64,25 @@ func GetUpdateAvailableForDevice(w http.ResponseWriter, r *http.Request) {
...
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
+		}
 		return
 	}
 	if _, ok := err.(*services.UpdateNotFoundError); ok {
 		err := errors.NewNotFound("Could not find update")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (biancarosa) [inline pkg/routes/devices.go:84]: Is there a way of using the logger that it's already injected on the context so we don't lose the requestID?

**Review by biancarosa — COMMENTED** [review]
Did minor comments about the use of the logger, but this is pretty great overall. Most of them are the same issue, which should be fairly easy to fix.

File: pkg/routes/imagesets.go (hunk 7 above)
Comment (Avielyo10) [reply pkg/routes/imagesets.go:255]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 18 above)
Comment (Avielyo10) [inline pkg/routes/devices.go:147]:
```suggestion
			s.log.WithField("error", result).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 14 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:129]:
```suggestion
		s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go [hunk 20]
```diff
@@ -96,13 +106,17 @@ func GetDeviceImageInfo(w http.ResponseWriter, r *http.Request) {
...
+			log.WithField("error", result).Error("Error while trying to encode")
+		}
 		return
 	}
 	if _, ok := err.(*services.DeviceNotFoundError); ok {
 		err := errors.NewNotFound("Could not find device")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/devices.go:118]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 13 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:110]:
```suggestion
			s.log.WithField("error", result).Error("Error while trying to encode")
```

File: pkg/routes/devices.go [hunk 21]
```diff
@@ -83,7 +91,9 @@ func GetUpdateAvailableForDevice(w http.ResponseWriter, r *http.Request) {
 		"statusCode": apierr.GetStatus(),
 		"error":      apierr.Error(),
 	}).Error("Error retrieving updates for device")
-	json.NewEncoder(w).Encode(&err)
+	if err := json.NewEncoder(w).Encode(&err); err != nil {
+		log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/devices.go:95]:
```suggestion
		s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 12 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:68]:
```suggestion
			s.log.WithField("error", result).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 11 above)
Comment (Avielyo10) [inline pkg/routes/devices.go:76]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 19 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:84]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 18 above)
Comment (Avielyo10) [inline pkg/routes/devices.go:147]:
```suggestion
			s.log.WithField("error", result).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 16 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:155]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 17 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:174]:
```suggestion
		s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 22]
```diff
@@ -93,7 +93,9 @@ func ImageSetCtx(next http.Handler) http.Handler {
 			}).Error("Error retrieving account")
 			err := errors.NewBadRequest(err.Error())
 			w.WriteHeader(err.GetStatus())
-			json.NewEncoder(w).Encode(&err)
+			if err := json.NewEncoder(w).Encode(&err); err != nil {
+				log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:97]:
```suggestion
				s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 23]
```diff
@@ -102,15 +104,19 @@ func ImageSetCtx(next http.Handler) http.Handler {
 			if err != nil {
 				err := errors.NewBadRequest(err.Error())
 				w.WriteHeader(err.GetStatus())
-				json.NewEncoder(w).Encode(&err)
+				if err := json.NewEncoder(w).Encode(&err); err != nil {
+					log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:108]:
```suggestion
					s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 24]
```diff
@@ -102,15 +104,19 @@ func ImageSetCtx(next http.Handler) http.Handler {
...
 				return
 			}
 			result := db.DB.Where("account = ? and Image_sets.id = ?", account, imageSetID).Find(&imageSet)
 
 			if result.Error != nil {
 				err := errors.NewNotFound(result.Error.Error())
 				w.WriteHeader(err.GetStatus())
-				json.NewEncoder(w).Encode(&err)
+				if err := json.NewEncoder(w).Encode(&err); err != nil {
+					log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:118]:
```suggestion
					s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 25]
```diff
@@ -144,7 +150,9 @@ func ListAllImageSets(w http.ResponseWriter, r *http.Request) {
 		s.Log.WithField("error", err.Error()).Error("Error listing all image sets - bad request")
 		err := errors.NewBadRequest(err.Error())
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:154]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 26]
```diff
@@ -155,7 +163,9 @@ func ListAllImageSets(w http.ResponseWriter, r *http.Request) {
 		s.Log.WithField("error", countResult.Error.Error()).Error("Error counting results for image sets list")
 		countErr := errors.NewInternalServerError()
 		w.WriteHeader(countErr.GetStatus())
-		json.NewEncoder(w).Encode(&countErr)
+		if err := json.NewEncoder(w).Encode(&countErr); err != nil {
+			log.WithField("error", countErr.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:167]:
```suggestion
			s.log.WithField("error", countErr.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go [hunk 27]
```diff
@@ -176,7 +186,9 @@ func ListAllImageSets(w http.ResponseWriter, r *http.Request) {
 		s.Log.WithField("error", countResult.Error.Error()).Error("Image sets not found")
 		err := errors.NewBadRequest("Not Found")
 		w.WriteHeader(err.GetStatus())
-		json.NewEncoder(w).Encode(&err)
+		if err := json.NewEncoder(w).Encode(&err); err != nil {
+			log.WithField("error", err.Error()).Error("Error while trying to encode")
```
Comment (Avielyo10) [inline pkg/routes/imagesets.go:190]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go (hunk 8 above)
Comment (Avielyo10) [reply pkg/routes/imagesets.go:265]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go (hunk 9 above)
Comment (Avielyo10) [reply pkg/routes/imagesets.go:277]:
```suggestion
			s.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/imagesets.go (hunk 6 above)
Comment (Avielyo10) [reply pkg/routes/imagesets.go:328]:
```suggestion
			s.log.WithField("error", errs).Error("Error while trying to encode")
```

File: pkg/routes/inventory.go (hunk 5 above)
Comment (Avielyo10) [reply pkg/routes/inventory.go:69]:
```suggestion
		 client.log.WithField("error", err.Error()).Error("Error while trying to encode")
```

File: pkg/routes/devices.go (hunk 10 above)
Comment (Avielyo10) [reply pkg/routes/devices.go:47]:
```suggestion

				log.WithField("error", err.Error()).Error("Error while trying to encode")
```

**Review by biancarosa — APPROVED** [review]

**Pull request merged**